mllam_data_prep/derived_variables.py

```diff
@@ -0,0 +1,478 @@
+    -------
+    hour_of_day_cos: Union[xr.DataArray, float]
+        cosine of the hour of day
+    hour_of_day_sin: Union[xr.DataArray, float]
+        sine of the hour of day
+    """
+    logger.info("Calculating hour of day")
+
+    # Get the hour of the day
+    if isinstance(time, xr.DataArray):
+        hour_of_day = time.dt.hour
+    elif isinstance(time, datetime.datetime):
+        hour_of_day = time.hour
+    else:
+        raise TypeError(
+            "Expected an instance of xr.DataArray or datetime object,"
+            f" but got {type(time)}."
+        )
+
+    # Cyclic encoding of hour of day
+    hour_of_day_cos, hour_of_day_sin = cyclic_encoding(hour_of_day, 24)
+
+    if isinstance(hour_of_day_cos, xr.DataArray):
+        # Add attributes
+        hour_of_day_cos.name = "hour_of_day_cos"
+        hour_of_day_cos.attrs[
+            "long_name"
+        ] = "Cosine component of cyclically encoded hour of day"
+        hour_of_day_cos.attrs["units"] = "1"
+
+    if isinstance(hour_of_day_sin, xr.DataArray):
+        # Add attributes
+        hour_of_day_sin.name = "hour_of_day_sin"
+        hour_of_day_sin.attrs[
+            "long_name"
+        ] = "Sine component of cyclically encoded hour of day"
+        hour_of_day_sin.attrs["units"] = "1"
+
+    return hour_of_day_cos, hour_of_day_sin
+
+
+def calculate_day_of_year(time):
+    """
+    Function for calculating day of year features with a cyclic encoding
+
+    Parameters
+    ----------
+    time : Union[xr.DataArray, datetime.datetime]
+        Time
+
+    Returns
+    -------
+    day_of_year_cos: Union[xr.DataArray, float]
+        cosine of the day of year
+    day_of_year_sin: Union[xr.DataArray, float]
+        sine of the day of year
+    """
+    logger.info("Calculating day of year")
+
+    # Get the day of year
+    if isinstance(time, xr.DataArray):
+        day_of_year = time.dt.dayofyear
+    elif isinstance(time, datetime.datetime):
+        day_of_year = time.timetuple().tm_yday
+    else:
+        raise TypeError(
+            "Expected an instance of xr.DataArray or datetime object,"
+            f" but got {type(time)}."
+        )
+
+    # Cyclic encoding of day of year - use 366 to include leap years!
+    day_of_year_cos, day_of_year_sin = cyclic_encoding(day_of_year, 366)
+
+    if isinstance(day_of_year_cos, xr.DataArray):
+        # Add attributes
+        day_of_year_cos.name = "day_of_year_cos"
+        day_of_year_cos.attrs[
+            "long_name"
+        ] = "Cosine component of cyclically encoded day of year"
+        day_of_year_cos.attrs["units"] = "1"
+
+    if isinstance(day_of_year_sin, xr.DataArray):
+        # Add attributes
+        day_of_year_sin.name = "day_of_year_sin"
+        day_of_year_sin.attrs[
+            "long_name"
+        ] = "Sine component of cyclically encoded day of year"
+        day_of_year_sin.attrs["units"] = "1"
+
+    return day_of_year_cos, day_of_year_sin
+
+
+def cyclic_encoding(data, data_max):
```

Comment:
I agree that it is getting long and it would be nice to split it up. The main reason for not adding it to `mllam_data_prep.ops` is that I don't know what "ops" stands for and what can/should be included here.

The idea and structure makes sense but the implementation is not very clear, probably because I am not very familiar with object oriented programming.

Reply:
I have also been a bit puzzeled about what the "ops" stands for. I think I have interpreted is as "operations", but not sure if this is correct. Do we document the meaning somewhere? Otherwise, we could maybe consider to rename it to make it more self-explainable (for another PR then).